### PR TITLE
406 -  Fetching dataset data using getServerSideProps 

### LINKED ISSUE #406: Fetching dataset data using getServerSideProps
### Context
Currently, we fetch the dataset at runtime on the dataset page. We want to replace it with the server-side props.
 
### Problem or idea

We should update the data fetching logic to utilize `getServerSideProps` in `page/[dataset]` and handle the resource availability check at the page level. 

### Solution or next step
1. Replace the runtime data fetching with `useServerSideProps`
2. Clean up the `dataset` availability check in the sub-components

### PR DESCRIPTION
**Pull request by nozomione**
## Issue Number

Closes #406

Base branch: `nozomione/370-add-get-dataset-state-helper` in PR #373

## Purpose/Implementation Notes

I've replaced the runtime data fetching with `getServerSideProps` in `page/[dataset]` and adjusted the codebase accordingly.

Changes include: 
-  Used `getServerSideProps` in `page/[dataset]` and updated the sub-components
- Minor adjustments in `useDatasetManager`:
   - Use `getDatasetState` and the `ok` property (to align with the latest changes in `dev`) 
   - Added the `isMyDatasetId` method for an ID check
- Minor adjustments in `usePollDatasetStatus`:
  - Remove one of the `useEffect` block (not required)
  - Remove `response.ok` check as it's already handled in `getDataset`

## Types of changes

- Refactor (addresses code organization and design mentioned in corresponding issue)

## Functional tests

N/A

## Checklist

- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots

N/A

### CODE REVIEW

File: src/components/Dataset/DatasetPageHeader.js [hunk 1]
```diff
@@ -7,68 +8,46 @@ import { DatasetProcessingError } from './DatasetProcessingError'
...
-        <DatasetRegenerate dataset={dataset} />
-      </Block>
-    )
-  }
+  // finds the current dataset state
+  const currentState = Object.keys(statusHeaders).find(
+    (state) => datasetStates[state]
+  )
+  // gets the corresponding component for the current dataset state
+  const Component = currentState ? statusHeaders[currentState] : null
```
Comment (davidsmejia) [inline src/components/Dataset/DatasetPageHeader.js:32]:
```suggestion
  const Component = statusHeaders[currentState]
```

If currentState is evaluated on load then this can simply just access the key.

File: src/components/Download/DatasetSummary/index.js [hunk 2]
```diff
@@ -13,34 +13,36 @@ import { Row } from 'components/shared/Row'
 import { SpiecesRow } from './SpeciesRow'
 import { TotalRow } from './TotalRow'
 
-// returns the count of expriment by spcecies
-const getExperimentCountBySpecies = (data, experiments) => {
+// returns the number of expriments for each organism
```
Comment (davidsmejia) [inline src/components/Download/DatasetSummary/index.js:16]: This looks like it needs to merge in dev.

File: src/hooks/usePollDatasetStatus.js
Comment (davidsmejia) [inline src/hooks/usePollDatasetStatus.js:1]: Its not immediately obvious to me why these changes took place. Though I agree with the less strict check on clearing the timer.

File: src/components/Dataset/DatasetPageHeader.js [hunk 3]
```diff
@@ -7,68 +8,46 @@ import { DatasetProcessingError } from './DatasetProcessingError'
...
-    return (
-      <Block>
-        <DatasetProcessing dataset={dataset} />
-      </Block>
-    )
+  const [datasetStates, setDatasetStates] = useState({})
+
+  useEffect(() => {
+    setDatasetStates(getDatasetState(dataset))
+  }, [dataset])
```
[the remark this answers is not in the document]
Comment (nozomione) [reply src/components/Dataset/DatasetPageHeader.js]: The states and the `useEffect` block have been replaced with constants 👍

File: src/components/Dataset/DatasetPageHeader.js (hunk 1 above)
Comment (nozomione) [reply src/components/Dataset/DatasetPageHeader.js]: It has been updated and with the addition of a new component `DatasetNotProcessed` for `isNotProcessed` state 👍

File: src/components/Dataset/DatasetPageHeader.js [hunk 4]
```diff
@@ -7,68 +8,46 @@ import { DatasetProcessingError } from './DatasetProcessingError'
...
+  const currentState = Object.keys(statusHeaders).find(
+    (state) => datasetStates[state]
+  )
+  // gets the corresponding component for the current dataset state
+  const Component = currentState ? statusHeaders[currentState] : null
 
   return (
     <FixedContainer pad="none">
-      <Box>
+      {Component ? (
```
[the remark this answers is not in the document]
Comment (nozomione) [reply src/components/Dataset/DatasetPageHeader.js]: It has been updated and with the addition of a new component `DatasetNotProcessed` for `isNotProcessed` state 👍

**Comment by nozomione** [comment]
I've applied your feedback and this PR is ready for another look. Thank you David!

**Comment by nozomione** [comment]
I've applied all the feedback from our 1:1, and this PR is now ready for your review. Thank you David!

File: src/hooks/useDatasetManager.js [hunk 5]
```diff
@@ -136,18 +131,26 @@ export const useDatasetManager = () => {
     return response
   }
 
+  // checks if the given dataset ID is My dataset ID
+  const isMyDatasetId = (id) => id === datasetId
+
   // takes download options, and optional dataset ID and one-off experiment accession code
   const startProcessingDataset = async (
```
Comment (davidsmejia) [inline src/hooks/useDatasetManager.js:138]: We will create 2 new issues for cleaning up the communication between formik and useDatasetManager.

File: src/pages/download/index.js [hunk 6]
```diff
@@ -39,7 +39,7 @@ export const Download = () => {
   }, [isDownloadable])
 
   useEffect(() => {
-    setIsDownloadable(getTotalSamples(dataset.data) > 0)
+    setIsDownloadable(getTotalSamples(dataset?.data) > 0)
```
Comment (davidsmejia) [inline src/pages/download/index.js:42]:
```suggestion
    if (dataset) {
      setIsDownloadable(getTotalSamples(dataset.data) > 0)
    } else {
      setIsDownloadable(false)
    }
    
```

**Review by davidsmejia — COMMENTED** [review]
Left some comments here

**Comment by nozomione** [comment]
I've applied the feedback and this PR is ready for another look. Thank you David!

File: src/hooks/useDatasetManager.js [hunk 7]
```diff
@@ -113,21 +112,17 @@ export const useDatasetManager = () => {
...
 
-    if (!id && datasetId) {
+    if (ok && isMyDatasetId(id)) {
       setDataset(response)
     }
 
-    // removes this dataset ID from processingDatasets[] if it exists
-    if (!isProcessing && success !== null) {
+    // sets the error if any, otherwise resets it
+    setError(!ok ? statusCode : null)
```
Comment (davidsmejia) [inline src/hooks/useDatasetManager.js:121]:
```suggestion
    setError(ok ? null : statusCode)
```

**Review by davidsmejia — APPROVED** [review]
LGTM, 1 minor change suggested
